### PR TITLE
feat: hardfork proposal feature

### PR DESCRIPTION
**Pull request by driftluo**
## What this PR does / why we need it?

This PR provides the full process of hardfork, but does not include the initialization process from the user command line parameters

- ref: https://github.com/axonweb3/axon/discussions/1317

### What is the impact of this PR?

No Breaking Change

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [ ] E2E Tests
- [x] Web3 Compatible Tests
- [ ] OCT 1-5 And 12-15
- [ ] OCT 6-10
- [ ] OCT 11
- [ ] OCT 16-19
- [ ] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Chaos CI*                                | Test the liveness and robustness of Axon under terrible network condition |
| *Cargo Clippy*                            | Run `cargo clippy --all --all-targets --all-features`                     |
| *Coverage Test*                           | Get the unit test coverage report                                         |
| *E2E Test*                                | Run end-to-end test to check interfaces                                   |
| *Code Format*                             | Run `cargo +nightly fmt --all -- --check` and `cargo sort -gwc`           |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

File: devtools/chain/specs/multi_nodes/chain-spec.toml [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 [genesis]
 timestamp = 1680249207
-extra_data = []
+extra_data = { block_number = 0, hardforks = [] }
```
Comment (chaoticlonghair) [inline devtools/chain/specs/multi_nodes/chain-spec.toml]: Move to the last field of `genesis` section, then:
```suggestion
[genesis.extra_data]
block_number = 0
hardforks = []
```

File: core/consensus/src/engine.rs [hunk 2]
```diff
@@ -608,6 +641,19 @@ impl<Adapter: ConsensusAdapter + 'static> ConsensusEngine<Adapter> {
...
+            if self_proposal
+                .as_ref()
+                .map(|v| v == &data)
+                .unwrap_or_default()
+            {
+                // remove self hardfork proposal
+                self_proposal.take();
+            }
+        }
+
```
Comment (KaoImin) [inline core/consensus/src/engine.rs]: Whether to remove a `hardfork_info` from `node_info` should be determined by the `hardfork info` in metadata.

**Review by KaoImin — COMMENTED** [review]
Doesn't this PR include parsing the `hardfork` arguments from command line?

File: core/consensus/src/engine.rs (hunk 2 above)
Comment (driftluo) [reply core/consensus/src/engine.rs]: The generated block already contains hardfork information, then the proposal has been confirmed. At this time, the proposal for the next block does not need to be submitted for hardfork again.

**Comment by driftluo** [comment]
> Doesn't this PR include parsing the `hardfork` arguments from command line?

wait until Boyu's command line reconstruction is completed before submitting it.

File: core/consensus/src/engine.rs (hunk 2 above)
Comment (KaoImin) [reply core/consensus/src/engine.rs]: What I mean is that it is not rigorous to determine that hardfork info can be removed at the time of block generated, it should be after the info written to the metadata store. So the operation should be judgement by compare the hardfork info in metadata:

```rust
// These code does not test
let store_hardfork_info = self.adapter.get_hardfork_info(); 
let mut self_proposal = self.node_info.hardfork_proposals.write().unwrap();

if self_proposal == store_hardfork_info {
	self_proposal.take();
}
```

**Comment by Flouse** [comment]
### Unit Test
No space left on device - https://github.com/axonweb3/axon/actions/runs/6105563759/job/16569277591#step:5:1566

=> Clean the cache and rerun -> https://github.com/axonweb3/axon/actions/runs/6105563759 ✅

**Review by KaoImin — DISMISSED** [review]

**Comment by KaoImin** [comment]
> > Doesn't this PR include parsing the `hardfork` arguments from command line?
> 
> wait until Boyu's command line reconstruction is completed before submitting it.

@driftluo  #1402 has merged

**Review by chaoticlonghair — APPROVED** [review]

**Review by KaoImin — APPROVED** [review]

**Comment by chaoticlonghair** [comment]
Failed status duet to [a nondeterministic failure](https://github.com/axonweb3/axon/actions/runs/6143523464/job/16667140761#step:5:1852), I have met it many times in my local development environment.

**Pull request merged**